pkg/repository/provider/unified_repo.go

```diff
@@ -129,60 +237,108 @@ func (urp *unifiedRepoProvider) EnsureUnlockRepo(ctx context.Context, param Repo
+func (urp *unifiedRepoProvider) GetPassword(param interface{}) (string, error) {
+	repoParam, ok := param.(RepoParam)
+	if !ok {
+		return "", errors.New("invalid parameter")
 	}
 
-	buf, err := secretStore.Get(repokey.RepoKeySelector())
+	repoPassword, err := getRepoPassword(urp.credentialGetter.FromSecret, repoParam)
 	if err != nil {
-		return "", errors.Wrap(err, "error to get password buffer")
+		return "", errors.Wrap(err, "error to get repo password")
 	}
 
-	return strings.TrimSpace(string(buf)), nil
+	return repoPassword, nil
 }
 
-func (urp *unifiedRepoProvider) getRepoOption(param RepoParam) (udmrepo.RepoOptions, error) {
-	repoPassword, err := funcTable.getRepoPassword(urp.credentialGetter.FromSecret, param)
-	if err != nil {
-		return udmrepo.RepoOptions{}, errors.Wrap(err, "error to get repo password")
+func (urp *unifiedRepoProvider) GetStoreType(param interface{}) (string, error) {
+	repoParam, ok := param.(RepoParam)
+	if !ok {
+		return "", errors.New("invalid parameter")
 	}
 
-	storeVar, err := funcTable.getStorageVariables(param.BackupLocation, param.SubDir)
-	if err != nil {
-		return udmrepo.RepoOptions{}, errors.Wrap(err, "error to get storage variables")
+	return getStorageType(repoParam.BackupLocation), nil
+}
+
+func (urp *unifiedRepoProvider) GetStoreOptions(param interface{}) (map[string]string, error) {
+	repoParam, ok := param.(RepoParam)
+	if !ok {
+		return map[string]string{}, errors.New("invalid parameter")
 	}
 
-	storeCred, err := funcTable.getStorageCredentials(param.BackupLocation, urp.credentialGetter.FromFile)
+	storeVar, err := funcTable.getStorageVariables(repoParam.BackupLocation, repoParam.BackupRepo.Spec.VolumeNamespace)
 	if err != nil {
-		return udmrepo.RepoOptions{}, errors.Wrap(err, "error to get repo credentials")
+		return map[string]string{}, errors.Wrap(err, "error to get storage variables")
 	}
 
-	repoOption := udmrepo.RepoOptions{
-		StorageType:    getStorageType(param.BackupLocation),
-		RepoPassword:   repoPassword,
-		ConfigFilePath: getRepoConfigFile(urp.workPath, string(param.BackupLocation.UID)),
-		Ownership: udmrepo.OwnershipOptions{
-			Username:   ownership.GetRepositoryOwner().Username,
-			DomainName: ownership.GetRepositoryOwner().DomainName,
-		},
-		StorageOptions: make(map[string]string),
-		GeneralOptions: make(map[string]string),
+	storeCred, err := funcTable.getStorageCredentials(repoParam.BackupLocation, urp.credentialGetter.FromFile)
+	if err != nil {
+		return map[string]string{}, errors.Wrap(err, "error to get repo credentials")
 	}
 
+	storeOptions := make(map[string]string)
 	for k, v := range storeVar {
-		repoOption.StorageOptions[k] = v
+		storeOptions[k] = v
 	}
 
 	for k, v := range storeCred {
-		repoOption.StorageOptions[k] = v
+		storeOptions[k] = v
 	}
 
-	return repoOption, nil
+	return storeOptions, nil
+}
+
+func getRepoPassword(secretStore credentials.SecretStore, param RepoParam) (string, error) {
```

Comment:
could getRepoPassword be GetRepoPassword？I need reuse this function in other package

Reply:
`unified_repo.go` is an implementation of `repository.Provider` interface, I don't think it is a good idea to call any functions in this file without through the interface. Moreover, in any other places, the caller should not understand what a `RepoParam` is.

If the uploader wants to get the repo password, it can get it in the similar way of `getRepoPassword`, but the uploader doesn't have to use the same way necessarily. This means, it is reasonable for the uploader to have its own implementation.
Moreover, the logic of `getRepoPassword` is very little, so uploader doesn't have to share one function with unified_repo.go.

Reply:
ok, I'll implement it my own package